cpp/ql/src/Likely Bugs/Conversion/LossyFunctionResultCast.ql

```diff
@@ -1,18 +1,39 @@
 /**
  * @name Lossy function result cast
  * @description Finds function calls whose result type is a floating point type, and which are casted to an integral type.
- *              Includes only expressions with implicit cast and excludes function calls to ceil, floor and round. {This is a gcc check; doesn't seem wildly useful.}
+ *              Includes only expressions with implicit cast and excludes function calls to ceil, floor and round.
  * @kind problem
  * @id cpp/lossy-function-result-cast
  * @problem.severity warning
+ * @precision medium
+ * @tags correctness
  */
 import cpp
 
+predicate whitelist(string fName) {
+  fName = "ceil" or
+  fName = "ceilf" or
+  fName = "ceill" or
+  fName = "floor" or
+  fName = "floorf" or
+  fName = "floorl" or
+  fName = "nearbyint" or
+  fName = "nearbyintf" or
+  fName = "nearbyintl" or
+  fName = "rint" or
+  fName = "rintf" or
+  fName = "rintl" or
+  fName = "round" or
+  fName = "roundf" or
+  fName = "roundl" or
+  fName = "trunc" or
+  fName = "truncf" or
+  fName = "truncl"
```

Comment:
My main worry is that this list is that this list isn't exhaustive. It doesn't include user-defined wrappers around rounding functions, for example. I also notice that `pow(2, smallPositiveInteger)` is a common FP.

Reply:
I didn't find any cases of user-defined wrappers 'in the wild', but I've added this functionality.

I've also add `pow` (and variants) as an exception, provided the first number (base) has a compile time value that is a whole number.

Reply:
I ran the query again on a bunch of LGTM projects and found [this FP](https://lgtm.com/projects/g/davidstutz/superpixels-revisited/snapshot/1c646d416b0c0f07f3ef5a5231f2a07b74668fbe/files/lib_vlfeat/vl/mathop.h?sort=name&dir=ASC&mode=heatmap#L177), where `__builtin_ceil` is called. Apparently, GCC has builtin versions of some of these operations. AFAICT from [the compiler docs](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html), only the ceil and floor functions are available as builtins, but the [`mathop.h` file](https://github.com/vlfeat/vlfeat/blob/master/vl/mathop.h#L163) also uses `__builtin_lroundf` and apparently gets away with it. Perhaps, to be on the safe side, we should just whitelist all these function names with a `__builtin_` prefix as well.

Reply:
Thanks for investigating.  I've added an exception for `__builtin_*` as well.